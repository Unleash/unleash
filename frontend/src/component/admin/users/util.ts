export const modalStyles = {
    overlay: {
        position: 'absolute',
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        backgroundColor: 'rgba(0, 0, 0, 0.25)',
        zIndex: 5,
    },
    content: {
        width: '500px',
        maxWidth: '90%',
        margin: '0',
        top: '50%',
        left: '50%',
        right: 'auto',
        bottom: 'auto',
        transform: 'translate(-50%, -50%)',
    },
};
