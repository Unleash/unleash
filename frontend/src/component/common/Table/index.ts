export { SortableTableHeader } from './SortableTableHeader/SortableTableHeader';
export { TableBody, TableRow } from '@mui/material';
export { Table } from './Table/Table';
export { TableCell } from './TableCell/TableCell';
export { TablePlaceholder } from './TablePlaceholder/TablePlaceholder';
export { VirtualizedTable } from './VirtualizedTable/VirtualizedTable';
