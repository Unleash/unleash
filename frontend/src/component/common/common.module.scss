/** Select **/

.truncate {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
}

.fullwidth {
    width: 100%;
}

.sectionPadding {
    padding: 0 16px;
}

.horisontalScroll {
    overflow-x: scroll;
    -webkit-overflow-scrolling: touch;
}

.horisontalScroll::-webkit-scrollbar {
    display: none;
}

.listLink {
    color: #212121;
    text-decoration: none;
    font-weight: normal;
    display: block;
}

.listLink:hover {
    color: #000;
}

@media (max-width: 920px) {
    .hideLt920 {
        display: none;
    }
}

@media (max-width: 600px) {
    .hideLt600 {
        display: none;
    }
}

.dataTableHeader {
    display: flex;
    justify-content: space-between;
    align-items: center;

    .titleText {
        margin: 0;
        font-size: 20px;
        line-height: 24px;
    }

    .actions {
        flex-shrink: 0;
    }
}

.switchWithLabel {
    display: flex;

    .label {
        padding-right: 16px;
        line-height: 24px;
    }
    .switch {
        display: inline-block;
        padding-right: 8px;
    }
}

.emptyState {
    display: flex;
    align-items: center;
    flex-direction: column;
    justify-content: center;
    min-height: 200px;
}

.toggleName {
    color: #37474f !important;
    font-weight: 700;
}

.toolbar {
    position: relative;
    padding: 0 24px 16px 24px;
    text-align: center;
}

.toolbarButton {
    position: absolute;
    top: 56px;
    right: 24px;
    z-index: 2;
}

.error {
    color: #d50000;
}

.headerTitle {
    font-size: var(--h1-size);
    margin: 0;
}

.listItem {
    padding: 0;
}

.section {
    padding: 8px 16px 8px 16px;
}

.contentPadding {
    padding: var(--card-padding);
}

.contentSpacing > * {
    margin: 0.5rem 0;
}

.searchField {
    margin-bottom: 2rem;
    max-width: 400px;
}
