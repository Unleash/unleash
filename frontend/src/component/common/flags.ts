export const P = 'P';
export const C = 'C';
export const E = 'E';
export const EEA = 'EEA';
export const RE = 'RE';
export const SE = 'SE';
export const UG = 'UG';
