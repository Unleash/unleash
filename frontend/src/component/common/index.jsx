import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import {
    List,
    MenuItem,
    Icon,
    ListItem,
    ListItemText,
    ListItemAvatar,
    Button,
    Avatar,
    Typography,
} from '@mui/material';
import { Apps } from '@mui/icons-material';

import styles from './common.module.scss';
import { ConditionallyRender } from './ConditionallyRender/ConditionallyRender';

export { styles };

export const AppsLinkList = ({ apps }) => (
    <List>
        <ConditionallyRender
            condition={apps.length > 0}
            show={apps.map(({ appName, description, icon }) => (
                <ListItem key={appName} className={styles.listItem}>
                    <ListItemAvatar>
                        <Avatar>
                            <ConditionallyRender
                                key={`avatar_conditional_${appName}`}
                                condition={icon}
                                show={<Icon>{icon}</Icon>}
                                elseShow={<Apps />}
                            />
                        </Avatar>
                    </ListItemAvatar>
                    <ListItemText
                        primary={
                            <Link
                                to={`/applications/${encodeURIComponent(
                                    appName
                                )}`}
                                className={[
                                    styles.listLink,
                                    styles.truncate,
                                ].join(' ')}
                            >
                                {appName}
                            </Link>
                        }
                        secondary={description || 'No description'}
                    />
                </ListItem>
            ))}
        />
    </List>
);
AppsLinkList.propTypes = {
    apps: PropTypes.array.isRequired,
};

export const DataTableHeader = ({ title, actions }) => (
    <div className={styles.dataTableHeader}>
        <div className={styles.title}>
            <Typography variant="h2" className={styles.titleText}>
                {title}
            </Typography>
        </div>
        {actions && <div className={styles.actions}>{actions}</div>}
    </div>
);
DataTableHeader.propTypes = {
    title: PropTypes.string,
    actions: PropTypes.any,
};

export const FormButtons = ({
    submitText = 'Create',
    onCancel,
    primaryButtonTestId,
}) => (
    <div>
        <Button
            data-testid={primaryButtonTestId}
            type="submit"
            color="primary"
            variant="contained"
        >
            {submitText}
        </Button>
        &nbsp;
        <Button type="cancel" onClick={onCancel}>
            Cancel
        </Button>
    </div>
);
FormButtons.propTypes = {
    submitText: PropTypes.string,
    onCancel: PropTypes.func.isRequired,
    primaryButtonTestId: PropTypes.string,
};

export const IconLink = ({ url, icon: IconComponent }) => (
    <a
        href={url}
        target="_blank"
        rel="noreferrer"
        className="mdl-color-text--grey-600"
    >
        <IconComponent />
    </a>
);
IconLink.propTypes = {
    url: PropTypes.string,
    icon: PropTypes.object,
};

export const MenuItemWithIcon = React.forwardRef(
    ({ icon: IconComponent, label, disabled, ...menuItemProps }, ref) => (
        <MenuItem
            disabled={disabled}
            style={{ display: 'flex', alignItems: 'center' }}
            {...menuItemProps}
        >
            <IconComponent />
            {label}
        </MenuItem>
    )
);
MenuItemWithIcon.propTypes = {
    icon: PropTypes.object,
    label: PropTypes.string,
    disabled: PropTypes.bool,
};

const badNumbers = [NaN, Infinity, -Infinity];
export function calc(value, total, decimal) {
    if (
        typeof value !== 'number' ||
        typeof total !== 'number' ||
        typeof decimal !== 'number'
    ) {
        return null;
    }

    if (total === 0) {
        return 0;
    }

    badNumbers.forEach(number => {
        if ([value, total, decimal].indexOf(number) > -1) {
            return number;
        }
    });

    return ((value / total) * 100).toFixed(decimal);
}

export const selectStyles = {
    control: provided => ({
        ...provided,
        border: '1px solid #607d8b',
        boxShadow: '0',
        ':hover': {
            borderColor: '#607d8b',
            boxShadow: '0 0 0 1px #607d8b',
        },
    }),
};
