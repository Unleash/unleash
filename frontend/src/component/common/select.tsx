import React from 'react';
import {
    FormControl,
    InputLabel,
    MenuItem,
    Select,
    SelectChangeEvent,
} from '@mui/material';
import { SELECT_ITEM_ID } from 'utils/testIds';

export interface ISelectOption {
    key: string;
    title?: string;
    label?: string;
}
export interface ISelectMenuProps {
    name: string;
    id: string;
    value?: string;
    label?: string;
    options: ISelectOption[];
    style?: object;
    onChange?: (event: SelectChangeEvent, child: React.ReactNode) => void;
    disabled?: boolean;
    className?: string;
    classes?: any;
}

const SelectMenu: React.FC<ISelectMenuProps> = ({
    name,
    value = '',
    label = '',
    options,
    onChange,
    id,
    disabled = false,
    className,
    classes,
    ...rest
}) => {
    const renderSelectItems = () =>
        options.map(option => (
            <MenuItem
                key={option.key}
                value={option.key}
                title={option.title || ''}
                data-testid={`${SELECT_ITEM_ID}-${option.label}`}
            >
                {option.label}
            </MenuItem>
        ));

    return (
        <FormControl variant="outlined" size="small" classes={classes}>
            <InputLabel htmlFor={id}>{label}</InputLabel>
            <Select
                name={name}
                disabled={disabled}
                onChange={onChange}
                className={className}
                label={label}
                id={id}
                value={value}
                {...rest}
            >
                {renderSelectItems()}
            </Select>
        </FormControl>
    );
};

export default SelectMenu;
