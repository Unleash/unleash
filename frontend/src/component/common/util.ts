import { weightTypes } from '../feature/FeatureView/FeatureVariants/FeatureVariantsList/AddFeatureVariant/enums';
import { IUiConfig } from 'interfaces/uiConfig';
import { IRoute } from 'interfaces/route';
import { IFeatureVariant } from 'interfaces/featureToggle';
import { format, isValid } from 'date-fns';

export const filterByConfig = (config: IUiConfig) => (r: IRoute) => {
    if (r.flag) {
        // Check if the route's `flag` is enabled in IUiConfig.flags.
        const flags = config.flags as unknown as Record<string, boolean>;
        return Boolean(flags[r.flag]);
    }

    if (r.configFlag) {
        // Check if the route's `configFlag` is enabled in IUiConfig.
        return Boolean(config[r.configFlag]);
    }

    return true;
};

export const scrollToTop = () => {
    window.scrollTo(0, 0);
};

export const trim = (value: string): string => {
    if (value && value.trim) {
        return value.trim();
    } else {
        return value;
    }
};

export const parseDateValue = (value: string) => {
    const date = new Date(value);
    return format(date, 'yyyy-MM-dd') + 'T' + format(date, 'HH:mm');
};

export const parseValidDate = (value: string): Date | undefined => {
    const parsed = new Date(value);

    if (isValid(parsed)) {
        return parsed;
    }
};

export const calculateVariantWeight = (weight: number) => {
    return weight / 10.0;
};

export function updateWeight(variants: IFeatureVariant[], totalWeight: number) {
    if (variants.length === 0) {
        return [];
    }
    const variantMetadata = variants.reduce(
        ({ remainingPercentage, variableVariantCount }, variant) => {
            if (variant.weight && variant.weightType === weightTypes.FIX) {
                remainingPercentage -= Number(variant.weight);
            } else {
                variableVariantCount += 1;
            }
            return {
                remainingPercentage,
                variableVariantCount,
            };
        },
        { remainingPercentage: totalWeight, variableVariantCount: 0 }
    );

    const { remainingPercentage, variableVariantCount } = variantMetadata;

    if (remainingPercentage < 0) {
        throw new Error('The traffic distribution total must equal 100%');
    }

    if (!variableVariantCount) {
        throw new Error('There must be at least one variable variant');
    }

    const percentage = parseInt(
        String(remainingPercentage / variableVariantCount)
    );

    return variants.map(variant => {
        if (variant.weightType !== weightTypes.FIX) {
            variant.weight = percentage;
        }
        return variant;
    });
}

export const modalStyles = {
    overlay: {
        position: 'absolute',
        top: 0,
        left: 0,
        right: 0,
        bottom: 0,
        backgroundColor: 'rgba(0, 0, 0, 0.25)',
        zIndex: 5,
    },
    content: {
        width: '500px',
        maxWidth: '90%',
        margin: '0',
        top: '50%',
        left: '50%',
        right: 'auto',
        bottom: 'auto',
        transform: 'translate(-50%, -50%)',
    },
};
