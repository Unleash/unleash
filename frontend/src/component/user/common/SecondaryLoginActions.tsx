import { styled, Typography } from '@mui/material';
import { Link } from 'react-router-dom';

const StyledContainer = styled('div')(({ theme }) => ({
    margin: 'auto auto 0 auto',
    width: '230px',
    [theme.breakpoints.down('md')]: {
        marginTop: theme.spacing(2),
    },
}));

const StyledLink = styled(Link)(({ theme }) => ({
    fontWeight: 'bold',
    textAlign: 'center',
}));

const StyledTypography = styled(Typography)(({ theme }) => ({
    fontWeight: 'bold',
    marginBottom: theme.spacing(1),
}));

const StyledRef = styled('a')(({ theme }) => ({
    fontWeight: 'bold',
    textAlign: 'center',
}));

const SecondaryLoginActions = () => {
    return (
        <StyledContainer>
            <StyledLink to="/forgotten-password">
                <StyledTypography variant="body2">
                    Forgot password?
                </StyledTypography>
            </StyledLink>
            <Typography variant="body2">
                Don't have an account?{' '}
                <StyledRef
                    href="https://www.getunleash.io/plans"
                    target="_blank"
                    rel="noopener noreferrer"
                >
                    Sign up
                </StyledRef>
            </Typography>
        </StyledContainer>
    );
};

export default SecondaryLoginActions;
