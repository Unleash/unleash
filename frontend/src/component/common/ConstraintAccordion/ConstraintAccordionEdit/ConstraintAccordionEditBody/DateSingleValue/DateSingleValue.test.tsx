import { parseDateValue } from 'component/common/util';

test(`Date component is able to parse midnight when it's 00`, () => {
    let f = parseDateValue('2022-03-15T12:27');
    let midnight = parseDateValue('2022-03-15T00:27');
    expect(f).toEqual('2022-03-15T12:27');
    expect(midnight).toEqual('2022-03-15T00:27');
});

test(`Date component - snapshot matching`, () => {
    let midnight = '2022-03-15T00:00';
    let midday = '2022-03-15T12:00';
    let obj = {
        midnight: parseDateValue(midnight),
        midday: parseDateValue(midday),
    };
    expect(obj).toMatchSnapshot();
});
