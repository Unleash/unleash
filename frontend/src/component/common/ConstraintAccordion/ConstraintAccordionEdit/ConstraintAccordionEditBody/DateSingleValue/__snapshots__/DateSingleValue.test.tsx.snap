// Vitest Snapshot v1

exports[`Date component - snapshot matching 1`] = `
{
  "midday": "2022-03-15T12:00",
  "midnight": "2022-03-15T00:00",
}
`;
